compilers/concrete-optimizer/concrete-optimizer/src/optimization/dag/multi_parameters/analyze.rs

```diff
@@ -1,38 +1,412 @@
+        let mut varianced = Self {
+            dag,
+            partitions,
+            variances,
+        };
+
+        // We loop, forwarding the noise, until it settles.
+        for forward_count in 0..MAX_FORWARDING {
+            // The noise gets gets computed from inputs down to outputs.
+            if varianced.forward_noise() {
+                break;
+            }
+            // If the noise is forwarded for the first time, we check that the dag is composable.
+            if forward_count == 0 {
+                varianced.check_composability()?;
+            }
+            // The noise of the inputs gets updated following the composition rules
+            varianced.apply_composition_rules();
+
+            assert!(
+                forward_count != MAX_FORWARDING,
+                "Forwarding of noise did not reach a fixed point."
+            );
+        }
+
+        // We return the varianced dag.
+        Ok(varianced)
+    }
+
+    #[allow(unused)]
+    fn get_operators_iter(&self) -> impl Iterator<Item = VariancedDagOperator<'_>> {
+        self.dag.get_indices_iter().map(|i| self.get_operator(i))
+    }
+
+    fn get_operators_iter_mut(&mut self) -> impl Iterator<Item = VariancedDagOperatorMut<'_>> {
+        struct Iter<'a> {
+            dag: &'a mut VariancedDag,
+            range: Range<usize>,
+        }
+
+        impl<'a> Iterator for Iter<'a> {
+            type Item = VariancedDagOperatorMut<'a>;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                let Range { start, end } = self.range;
+                if start == end {
+                    return None;
+                }
+                self.range = (start + 1)..end;
+                unsafe {
+                    Some(std::mem::transmute::<
+                        VariancedDagOperatorMut<'_>,
+                        VariancedDagOperatorMut<'a>,
+                    >(
+                        self.dag.get_operator_mut(OperatorIndex(start))
+                    ))
+                }
+            }
+        }
+        Iter {
+            range: 0..self.dag.len(),
+            dag: self,
+        }
+    }
+
+    #[allow(unused)]
+    fn get_composed_inputs_iter(&self) -> impl Iterator<Item = ComposedInput<'_>> {
+        self.dag
+            .composition
+            .clone()
+            .into_iter()
+            .map(move |(id, from)| ComposedInput {
+                dag: self,
+                id,
+                from,
+            })
+    }
+
+    fn get_composed_inputs_iter_mut(&mut self) -> impl Iterator<Item = ComposedInputMut<'_>> {
+        struct Iter<'a> {
+            dag: &'a mut VariancedDag,
+            iter: std::collections::hash_map::IntoIter<OperatorIndex, Vec<OperatorIndex>>,
+        }
+
+        impl<'a> Iterator for Iter<'a> {
+            type Item = ComposedInputMut<'a>;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                self.iter.next().map(|(to, from)| unsafe {
```

Comment:
unsafe ? is this a rust limitation here or a design issue ?

Reply:
A bit of both to be honest. A rust limitation because you mostly can't provide an iterator over mutable references without unsafe, but there is also a design issue around the way I implemented it. I'll try to rethink this part of the code and come with something better suited.